searchlib/src/vespa/searchlib/features/closest_feature.cpp

```diff
@@ -0,0 +1,288 @@
+bool
+ClosestBlueprint::setup(const fef::IIndexEnvironment & env, const fef::ParameterList & params)
+{
+    if (params.size() < 1 || params.size() > 2) {
+        LOG(error, "%s: Wrong number of parameters, was %d, must be 1 or 2", getName().c_str(), (int) params.size());
+        return false;
+    }
+    _field_name = params[0].getValue();
+    if (params.size() == 2) {
+        _item_label = params[1].getValue();
+    }
+    auto fi = env.getFieldByName(_field_name);
+    if (fi == nullptr) {
+        LOG(error, "%s: Unknown field %s", getName().c_str(), _field_name.c_str());
+        return false;
+    }
+    auto dt = fi->get_data_type();
+    auto ct = fi->collection();
+    if (dt != search::index::schema::DataType::TENSOR ||
+        ct != search::index::schema::CollectionType::SINGLE) {
+        LOG(error, "%s: Field %s is not a single value tensor field", getName().c_str(), _field_name.c_str());
+        return false;
+    }
+    if (!fi->hasAttribute()) {
+        LOG(error, "%s: Field %s does not have an attribute", getName().c_str(), _field_name.c_str());
+        return false;
+    }
+    vespalib::string attr_type_spec = type::Attribute::lookup(env.getProperties(), _field_name);
+    if (attr_type_spec.empty()) {
+        LOG(error, "%s: Field %s lacks a type in index properties", getName().c_str(), _field_name.c_str());
+        return false;
+    }
+    _field_tensor_type = ValueType::from_spec(attr_type_spec);
+    if (_field_tensor_type.is_error() || _field_tensor_type.is_double() || _field_tensor_type.count_mapped_dimensions() == 0) {
+        LOG(error, "%s: Field %s invalid type: '%s'", getName().c_str(), _field_name.c_str(), attr_type_spec.c_str());
+        return false;
+    }
+    _output_tensor_type = ValueType::make_type(_field_tensor_type.cell_type(), _field_tensor_type.mapped_dimensions());
+    assert(!_output_tensor_type.is_double());
+    FeatureType output_type = FeatureType::object(_output_tensor_type);
+    describeOutput("out", "The closest tensor subspace.", output_type);
+    _field_id = fi->id();
+    _empty_output = vespalib::eval::value_from_spec(_output_tensor_type.to_spec(), FastValueBuilderFactory::get());
+    setup_identity_cells(_output_tensor_type, _identity_space, _identity_cells);
+    return true;
+}
+
+void
+ClosestBlueprint::prepareSharedState(const fef::IQueryEnvironment& env, fef::IObjectStore& store) const
+{
+    if (_item_label.has_value()) {
+        DistanceCalculatorBundle::prepare_shared_state(env, store, _item_label.value(), "closest");
+    } else {
+        DistanceCalculatorBundle::prepare_shared_state(env, store, _field_id, "closest");
+    }
+}
+
+fef::FeatureExecutor&
+ClosestBlueprint::createExecutor(const fef::IQueryEnvironment &env, vespalib::Stash &stash) const
+{
+    auto bundle = _item_label.has_value() ? DistanceCalculatorBundle(env, _field_id, _item_label.value(), "closest") : DistanceCalculatorBundle(env, _field_id, "closest");
+    if (bundle.elements().empty()) {
```

Comment:
If one of the calculators in the bundle is nullptr, we should also shortcut with an empty tensor.